components/service/glean/src/main/java/mozilla/components/service/glean/scheduler/PingScheduler.kt

```diff
@@ -0,0 +1,164 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.glean.scheduler
+
+import android.arch.lifecycle.Lifecycle
+import android.arch.lifecycle.LifecycleObserver
+import android.arch.lifecycle.OnLifecycleEvent
+import android.content.Context
+import android.support.annotation.VisibleForTesting
+import androidx.work.BackoffPolicy
+import androidx.work.Constraints
+import androidx.work.Data
+import androidx.work.NetworkType
+import androidx.work.OneTimeWorkRequestBuilder
+import androidx.work.WorkManager
+import mozilla.components.service.glean.Glean
+import mozilla.components.service.glean.GleanMetrics.GleanBaseline
+import mozilla.components.service.glean.ping.BaselinePing
+import mozilla.components.service.glean.ping.PingMaker
+import mozilla.components.service.glean.storages.PingStorageEngine
+import mozilla.components.service.glean.storages.StorageEngineManager
+import mozilla.components.support.base.log.logger.Logger
+import java.util.UUID
+import java.util.concurrent.TimeUnit
+
+/**
+ * This class is responsible for scheduling of glean pings. In order to do this in association
+ * with lifecycle events, it implements [LifecycleObserver].  On entering the background
+ * [Lifecycle.Event.ON_STOP] state, it will use [WorkManager] to schedule a [PingUploadWorker] task
+ * which is responsible for uploading the ping.
+ */
+internal class PingScheduler(
+    applicationContext: Context,
+    storageEngineManager: StorageEngineManager
+) : LifecycleObserver {
```

Comment:
It made sense to have them internal to the `PingScheduler` since that was the only place most of them were used, aside from the `PingUploadWorker` needing access to the `pingStorageEngine` in order to process the files.  I'm open to moving them back if it turns out they are better suited to live somewhere else.

Reply:
What I mean is we can inject them as dependencies through the constructor, making testing easier, rather than relying on interna, e.g. `PingScheduler(context, StorageEngineManager(), PingMaker(), ...)`

Reply:
Ahh yes, I see where you are going now.  Yes, absolutely.